### PR TITLE
Refactor shard transfer logic

### PR DESCRIPTION
**Pull request by timvisee**
Tracked in <https://github.com/qdrant/qdrant/issues/2432>.

Depends on <https://github.com/qdrant/qdrant/pull/2989>.

Relatively basic refactoring of our shard transfer code into the following modules:

- `shard/transfer/mod.rs`: type and trait definitions
- `shard/transfer/driver.rs`: functions driving shard transfer from consensus
- `shard/transfer/helpers.rs`: helper functions for validation, conflict checking, ...
- `shard/transfer/snapshot.rs`: shard snapshot transfer logic
- `shard/transfer/stream_records.rs`: shard streaming records transfer logic
- `shard/transfer/transfer_tasks_pool.rs`: pool implementation to keep track of transfers

Contains no logic changes.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### CODE REVIEW

File: lib/collection/src/shards/transfer/mod.rs [hunk 1]
```diff
@@ -24,6 +28,54 @@ const CONSENSUS_CONFIRM_RETRY_DELAY: Duration = Duration::from_secs(1);
...
+
+impl ShardTransfer {
+    pub fn key(&self) -> ShardTransferKey {
+        ShardTransferKey {
+            shard_id: self.shard_id,
+            from: self.from,
+            to: self.to,
+        }
+    }
+}
```
Comment (ffuugoo) [inline lib/collection/src/shards/transfer/mod.rs:66]: Can we, please, move this right after `struct ShardTransfer`? 🙏

File: lib/collection/src/shards/transfer/snapshot.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+/// After this function, the following will happen:
+///
+/// - The local shard is un-proxified
+/// - The shard transfer is finished
+/// - The remote shard state is set to `Active` through consensus
+///
+/// # Diagram
+///
+/// Here's a rough sequence diagram for the shard snasphot transfer process with the consensus,
+/// sender and receiver actors:
```
Comment (ffuugoo) [inline lib/collection/src/shards/transfer/snapshot.rs:75]: 🤯🤩🎉

**Review by ffuugoo — APPROVED** [review]
LGTM! __Love__ `transfer_snapshot` diagram. Thanks, @timvisee. 🙌

File: lib/collection/src/shards/transfer/mod.rs (hunk 1 above)
Comment (ffuugoo) [reply lib/collection/src/shards/transfer/mod.rs]: Thanks! 🤝

**Comment by timvisee** [comment]
> LGTM! **Love** `transfer_snapshot` diagram. Thanks, @timvisee. 🙌

:smile: :tada: 

It might also be the last one though. It was a pain to _draw_.

**Comment by ffuugoo** [comment]
> Thanks! It might also be the last one though. It was a pain to _draw_.

We need markdown inline images support in rustdoc (and in editors). All those _beautiful_ diagrams are _wasted_ on @generall's iPad.

**Pull request merged**